onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc

```diff
@@ -140,34 +174,39 @@ static size_t GetPaddedByteSize(size_t size) {
 }
 
 void ModelBuilder::GetTargetDevices() {
-  // GetTargetDevices is only supported on API 29+
-  if (GetAndroidSdkVer() < 29)
+  if (target_device_option_ == TargetDeviceOption::ALL_DEVICES)
     return;
 
-  if (target_device_option_ == TargetDeviceOption::ALL_DEVICES)
+  nnapi_target_devices_ = GetTargetDevicesHelper(nnapi_, target_device_option_, GetAndroidSdkVer());
+}
+
+void ModelBuilder::LogOperationsSupportedByTargetDevices() {
+  if (!log_op_target_device_support_)
     return;
 
-  const std::string nnapi_cpu("nnapi-reference");
-  uint32_t num_devices = 0;
-  THROW_ON_ERROR_WITH_NOTE(nnapi_->ANeuralNetworks_getDeviceCount(&num_devices),
-                           "Getting count of available devices");
+  // getSupportedOperationsForDevices is only supported on API 29+
+  if (GetAndroidSdkVer() < 29)
+    return;
 
-  for (uint32_t i = 0; i < num_devices; i++) {
-    ANeuralNetworksDevice* device = nullptr;
-    const char* device_name = nullptr;
-    THROW_ON_ERROR_WITH_NOTE(nnapi_->ANeuralNetworks_getDevice(i, &device),
-                             "Getting " + std::to_string(i) + "th device");
+  if (nnapi_target_devices_.empty())
+    return;
 
-    THROW_ON_ERROR_WITH_NOTE(nnapi_->ANeuralNetworksDevice_getName(device, &device_name),
-                             "Getting " + std::to_string(i) + "th device's name");
+  const auto nnapi_model_size = nnapi_operations_.size();
+  std::unique_ptr<bool[]> nnapi_ops_support_flags(new bool[nnapi_model_size]);
 
-    bool device_is_cpu = nnapi_cpu == device_name;
-    if ((target_device_option_ == TargetDeviceOption::CPU_DISABLED && !device_is_cpu) ||
-        (target_device_option_ == TargetDeviceOption::CPU_ONLY && device_is_cpu)) {
-      nnapi_target_devices_.push_back(device);
-      LOGS_DEFAULT(VERBOSE) << "Target device [" << device_name << "] added";
-    }
+  THROW_ON_ERROR_WITH_NOTE(
+      nnapi_->ANeuralNetworksModel_getSupportedOperationsForDevices(
+          nnapi_model_->model_, nnapi_target_devices_.data(), nnapi_target_devices_.size(),
+          nnapi_ops_support_flags.get()),
+      "Checking supported operations for devices");
+
+  std::string out = "";
+  for (size_t i = 0; i < nnapi_model_size; i++) {
+    out += "[" + std::to_string(i) + ", " + std::to_string(nnapi_operations_[i]) +
+           ", " + std::to_string(nnapi_ops_support_flags[i]) + "],";
   }
```

Comment:
Move this into a lambda that's only called in LOGS_DEFAULT so you're not formatting up a string you're not going to use. std::ostringstream is probably a better choice to create the string as well.

Reply:
Removed

---
In reply to: [454726022](https://github.com/microsoft/onnxruntime/pull/4499#discussion_r454726022) [](ancestors = 454726022)